assets/js/components/user-input/UserInputPreview.js

```diff
@@ -58,7 +58,12 @@ export default function UserInputPreview( { back, goTo } ) {
 		setIsNavigating( true );
 		const response = await saveUserInputSettings();
 		if ( ! response.error ) {
-			global.location.assign( addQueryArgs( dashboardURL, { notification: 'user_input_success' } ) );
+			let [ sanitizedRedirectURL, hash ] = ( redirectURL ?? dashboardURL ).split( '#' );
+			sanitizedRedirectURL = addQueryArgs( sanitizedRedirectURL, { notification: 'user_input_success' } );
+			if ( hash ) {
+				sanitizedRedirectURL += `#${ hash }`;
+			}
```

Comment:
As you can see [here](https://github.com/WordPress/gutenberg/blob/master/packages/url/src/add-query-args.js#L23), the `addQueryArgs` does not care about the `hash` at all. 
It takes the url, appends query arg to the url. 
What happens in the case of hash is this:
if 
```js
url = http://example.com/path?argument=value#hash
```
then 
```js
addQueryArgs(url, {test: "test"}) = "http://example.com/path?argument=value%23hash&test=test"
```

The `#` gets sandwiched into a query arg itself. What do you suggest I do here? Splitting seems to be the best thing as we need to get rid of the hash before adding query args to it and then add the hash later.

Reply:
I see what you mean @kostyalmm . It works fine when the hash is before the query string but is not handled when it is after which seems to be the correct place when both are present 🤔 

This seems to be a known bug in the package: https://github.com/WordPress/gutenberg/issues/16655

I feel like the safest way is to use `URL` rather than assume the hash is in a particular spot. This seems to be the direction the PR for that issue is taking as well.

For this specific case, I think we can avoid using `addQueryArgs` since we're simply adding a specific, known parameter which also doesn't require encoding.

```js
const url = new URL( redirectURL || dashboardURL );
// Here we don't use `addQueryArgs` due to a bug with how it handles hashes
// See https://github.com/WordPress/gutenberg/issues/16655
url.searchParams.set( 'notification', 'user_input_success' );
const finalURL = url.toString();
```